src/mlpack/bindings/cli/add_to_po.hpp

```diff
@@ -17,7 +17,9 @@
 #include <mlpack/core/util/is_std_vector.hpp>
 #include "map_parameter_name.hpp"
 
-#include <CLI/CLI.hpp>
+namespace CLI11 {
+  #include <CLI/CLI.hpp>
+}
```

Comment:
@robertohueso This method does not work, I removed the this code in the next commits

Reply:
Thanks for the update!

I was wondering, is [this](https://github.com/robertohueso/mlpack/commit/ef6fe2e46f9272443878f1a3f125a51375c9562f) helpful? I tested and it seems to work :)

Reply:
Hmm, interesting.
Also, I have already applied the same modification you have on `core`, but it was not able to find the `serialize` function in `neighbor_search` class. 
When I compiled directly your branch it was fine. However, when I applied your modifications on this branch I got a different error.

Reply:
Which error? Maybe I can help somehow.
Also, take into account this was just an extremely simple example to serialize a `bool` in just one method so this might a lot more complex. I'm still unsure on how to serialize a `boost::variant` :thinking: I'll have a look at that tomorrow.

Reply:
I am getting this type of error: 
```
/meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:150:56:   required from 'mlpack::bindings::cmd::CMDOption<N>::CMDOption(N, const string&, const string&, const string&, const string&, boo
l, bool, bool, const string&) [with N = mlpack::range::RSModel*; std::string = std::__cxx11::basic_string<char>]'
/meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:80:1:   required from here
/meta/mlpack/src/mlpack/bindings/cli/get_param.hpp:135:19: error: cannot convert 'mlpack::range::RSModel*' to 'mlpack::range::RSModel**' in assignment
  135 |   *((T**) output) = &GetParam<typename std::remove_pointer<T>::type>(
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  136 |       const_cast<util::ParamData&>(d));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Reply:
I belive this error is the error @rcurtin mention here: 
https://github.com/mlpack/mlpack/blob/cb982959d2aebe8cc3030546ee14fdd544173b57/src/mlpack/core/util/param.hpp#L918-L940

Reply:
Note that the error is a long and complex one, I only added a part of this error here.
I had this error with Boost when I tried to remove the version. of some files. However, in this case, I did not change the serialize function.

Reply:
I'm having issues to compile any of mlpack methods (at commit 28f99b3bb03a4cbb3365ef5abe0fcde4334efb46). This is probably because I'm not following the same compilation steps sequence as you, so what I'm doing right now is:
```shell
mkdir build
cd build
cmake ../
# either
make mlpack_range_search
# or
make mlpack_knn
```
Is this correct?

Also,
> https://github.com/mlpack/mlpack/blob/cb982959d2aebe8cc3030546ee14fdd544173b57/src/mlpack/core/util/param.hpp#L918-L940

I don't think this is the source of the error since you have not deleted `serialize(...)` function and, as far as I know, we only check for the existence of a function with that signature at compile time (but I could be totally wrong).

Reply:
@robertohueso You are following exactly the same sequence, you do not have to care about cross compiling, since it can not compile any way.

Reply:
I am having the same issue also since I have replaced boost program options by CLI11. 
Actually I am able to compile any method that does not have any `serialize` inside.
For example, I am able to compile PCA example without any issue, since there is not serialize inside.

Reply:
Ahh, that comment is actually inaccurate!  However probably not in a way that affects the discussion here.  https://github.com/mlpack/mlpack/pull/2449

Reply:
@rcurtin Happy to see you around :+1: :smile:

Reply:
Yep, just got back yesterday.  It was good to relax for a few days. :)

I was able to reproduce the issue in a2b18f9.  I haven't gotten to the bottom of it but I believe that `data::HasSerialize<>` is not working right for serializable classes anymore.  I know that at least the `int` parameter will need to be removed from all the `serialize()` methods.

Reply:
Ah, sorry, I was testing in a sloppy way.  It seems that making `RSModel::serialize()` take only one argument (i.e. removing the `const unsigned int`) is sufficient to make `mlpack_range_search` compile. :+1:

Reply:
I will test it tomorrow, but which one you have tested, those in `test_cereal` or the other in `mlpack/method`?
I just created test_cereal to make things easier.

Reply:
Oh, I think I did the same thing @robertohueso did---with revision `a2b18f9`, just `cmake ../` and `make mlpack_range_search`.  Maybe I didn't reproduce it correctly; if so, sorry about that.

Reply:
Hmm, that is interesting, when I am compiling without the version I am getting this type of error:
```
/meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:150:56:   required from 'mlpack::bindings::cmd::CMDOption<N>::CMDOption(N, const string&, const string&, const string&, const s[3174/9064]
l, bool, bool, const string&) [with N = mlpack::range::RSModel*; std::string = std::__cxx11::basic_string<char>]'
/meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:80:1:   required from here
/meta/mlpack/src/mlpack/bindings/cli/get_param.hpp:135:19: error: cannot convert 'mlpack::range::RSModel*' to 'mlpack::range::RSModel**' in assignment
  135 |   *((T**) output) = &GetParam<typename std::remove_pointer<T>::type>(
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  136 |       const_cast<util::ParamData&>(d));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:26,
                 from /meta/mlpack/src/mlpack/core/util/mlpack_main.hpp:37,
                 from /meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:16:
/meta/mlpack/src/mlpack/bindings/cli/get_raw_param.hpp: In instantiation of 'void mlpack::bindings::cmd::GetRawParam(const mlpack::util::ParamData&, const void*, void*) [with T = mlpack::ran
ge::RSModel*]':
/meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:151:59:   required from 'mlpack::bindings::cmd::CMDOption<N>::CMDOption(N, const string&, const string&, const string&, const string&, boo
l, bool, bool, const string&) [with N = mlpack::range::RSModel*; std::string = std::__cxx11::basic_string<char>]'
/meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:80:1:   required from here
/meta/mlpack/src/mlpack/bindings/cli/get_raw_param.hpp:85:19: error: cannot convert 'mlpack::range::RSModel*' to 'mlpack::range::RSModel**' in assignment
   85 |   *((T**) output) = &GetRawParam<typename std::remove_pointer<T>::type>(
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   86 |       const_cast<util::ParamData&>(d));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /meta/mlpack/src/mlpack/bindings/cli/default_param.hpp:93,
                 from /meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:21,
                 from /meta/mlpack/src/mlpack/core/util/mlpack_main.hpp:37,
                 from /meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:16:
/meta/mlpack/src/mlpack/bindings/cli/default_param_impl.hpp: In instantiation of 'std::string mlpack::bindings::cmd::DefaultParamImpl(const mlpack::util::ParamData&, const typename boost::di
sable_if<arma::is_arma_type<T1> >::type*, const typename boost::disable_if<mlpack::util::IsStdVector<T> >::type*, const typename boost::disable_if<mlpack::data::HasSerialize<T> >::type*, con
st typename boost::disable_if<std::is_same<T, std::__cxx11::basic_string<char> > >::type*, const typename boost::disable_if<std::is_same<T, std::tuple<mlpack::data::DatasetMapper<mlpack::dat
a::IncrementPolicy, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, arma::Mat<double> > > >::type*) [with T = mlpack::range::RSModel; std::string = std::__
cxx11::basic_string<char>; typename boost::disable_if<arma::is_arma_type<T1> >::type = void; typename boost::disable_if<mlpack::util::IsStdVector<T> >::type = void; typename boost::disable_i
f<mlpack::data::HasSerialize<T> >::type = void; typename boost::disable_if<std::is_same<T, std::__cxx11::basic_string<char> > >::type = void; typename boost::disable_if<std::is_same<T, std::
tuple<mlpack::data::DatasetMapper<mlpack::data::IncrementPolicy, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, arma::Mat<double> > > >::type = void]':
/meta/mlpack/src/mlpack/bindings/cli/default_param.hpp:85:68:   required from 'void mlpack::bindings::cmd::DefaultParam(const mlpack::util::ParamData&, const void*, void*) [with T = mlpack::
range::RSModel*]'
/meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:142:60:   required from 'mlpack::bindings::cmd::CMDOption<N>::CMDOption(N, const string&, const string&, const string&, const string&, boo
l, bool, bool, const string&) [with N = mlpack::range::RSModel*; std::string = std::__cxx11::basic_string<char>]'
/meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:80:1:   required from here
/meta/mlpack/src/mlpack/bindings/cli/default_param_impl.hpp:36:9: error: no match for 'operator<<' (operand types are 'std::ostringstream' {aka 'std::__cxx11::basic_ostringstream<char>'} and
 'mlpack::range::RSModel')
   36 |     oss << boost::any_cast<T>(data.value);
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /meta/mlpack/board/boost_1_73_0/boost/archive/xml_oarchive.hpp:19,
                 from /meta/mlpack/src/mlpack/core/data/has_serialize.hpp:18,
                 from /meta/mlpack/src/mlpack/prereqs.hpp:88,
                 from /meta/mlpack/src/mlpack/methods/range_search/range_search_main.cpp:14:
/usr/aarch64-linux-gnu/include/c++/10.1.0/ostream:108:7: note: candidate: 'std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::basic_ost
ream<_CharT, _Traits>::__ostream_type& (*)(
```